### PR TITLE
Parcelize: Implement missing FIR diagnostics

### PR DESCRIPTION
**Pull request by sfs**
This PR removes the dependency of the Parcelize diagnostics on the old backend and then implements the missing diagnostics for FIR. The latter required a small change in `AbstractDiagnosticCollectorVisitor`, which I've moved to a separate commit.

---

The original code contained a comment about how we should "abstract code from `ParcelSerializer` or `IrParcelSerializerFactory` to avoid duplication of allowed types checking". That's a valid concern, but ultimately introducing such an abstraction ends up being more code than duplicating the check. The check itself is simple and the list of allowed types can be shared between the FE10 and FIR implementation.

### CODE REVIEW

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/collectors/AbstractDiagnosticCollectorVisitor.kt [hunk 1]
```diff
@@ -334,6 +336,17 @@ abstract class AbstractDiagnosticCollectorVisitor(
         }
     }
 
+    private inline fun <R> withTypeRefAnnotationContainer(annotationContainer: FirTypeRef, block: () -> R): R {
+        var containingTypeRef = context.annotationContainers.lastOrNull()?.safeAs<FirResolvedTypeRef>()
```
Comment (demiurg906) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/collectors/AbstractDiagnosticCollectorVisitor.kt:340]: Please never user `safeAs`. This function was introduced for cases when `as?` stays in receiver position to avoid redundat parenthesis (`(x as? A)?.foo()` vs `x.safeAs<A>()?.foo()`). But this function is not understandable by compiler, so it won't report `USELESESS_CAST` or `CAST_NEVER_SUCCEED` or smth else if receiver of this call will be changed. Also It has worse IDE support comparing to regular `as?`

I'll fix this one by myself, so it's just note for future

File: plugins/parcelize/parcelize-compiler/testData/diagnostics/customParcelers.fir.kt [hunk 2]
```diff
@@ -19,27 +19,27 @@ class StringClassParceler : Parceler<String> {
...
 
 @Parcelize
 class Test(
-    val a: @WriteWith<StringParceler> Int,
+    val a: @WriteWith<<!PARCELER_TYPE_INCOMPATIBLE!>StringParceler<!>> Int,
     val b: @WriteWith<StringParceler> String,
-    val c: @WriteWith<StringParceler> CharSequence,
-    val d: @WriteWith<CharSequenceParceler> String,
+    val c: @WriteWith<<!PARCELER_TYPE_INCOMPATIBLE!>StringParceler<!>> CharSequence,
+    val d: @WriteWith<<!PARCELER_TYPE_INCOMPATIBLE!>CharSequenceParceler<!>> String,
```
Comment (demiurg906) [inline plugins/parcelize/parcelize-compiler/testData/diagnostics/customParcelers.fir.kt:34]: Is it ok that K1 implementation does not report `PARCELER_TYPE_INCOMPATIBLE` on `d`?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/collectors/AbstractDiagnosticCollectorVisitor.kt (hunk 1 above)
Comment (sfs) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/collectors/AbstractDiagnosticCollectorVisitor.kt:340]: Ack

File: plugins/parcelize/parcelize-compiler/testData/diagnostics/customParcelers.fir.kt (hunk 2 above)
Comment (sfs) [reply plugins/parcelize/parcelize-compiler/testData/diagnostics/customParcelers.fir.kt:34]: It's going to be a separate PR/issue, since the check in K1 for `WriteWith` is wrong, imo. Or rather, it is explicitly checking that the given parceler can be used to *write* the annotated type, but the generated code will also use it to try and read the same type.

More precisely, in K1 we handle `@WriteWith<P> T1` by finding a `Parceler<T2>` superclass of `P` and then testing that `T1` is a subtype of `T2`. In this example that succeeds for `d` since `T2` is `CharSequence` and `T1` is `String`. If instead we want to check that we can both read and write the annotated type we should check that `P` is a subtype of `Parceler<T1>`, which is what the FIR checker does.

Bottom line is I think the existing check is wrong and I didn't want to reproduce the buggy behavior in K2. I'll change it to match if it turns out that I'm wrong and that there's a legitimate use case here...

---

Incidentally, the same problem exists for `TypeParceler`, but that's just in the form of a variance annotation in the definition of the annotation.

Comment (demiurg906) [reply plugins/parcelize/parcelize-compiler/testData/diagnostics/customParcelers.fir.kt:34]: Got it, thanks. In that case, I'll merge PR after CI build will pass
